### PR TITLE
Fix debug logs failing with persistent connection

### LINKED ISSUE #33047: Network module using persistent connection fails when debug logs are enabled
##### ISSUE TYPE

 - Bug Report

##### COMPONENT NAME

ios_config
junos_config

##### ANSIBLE VERSION

```
2.5
```

##### STEPS TO REPRODUCE

Enable debug logs

```yaml
---
-  name: test
   hosts: vsrx
   connection: local
   gather_facts: no

   tasks:
   - name: test
     junos_command:
          commands:
            - show version
     register: response
```
OR
```yaml
---
-  name: test
   hosts: vsrx
   connection: netconf
   gather_facts: no

   tasks:
   - name: test
     junos_command:
          commands:
            - show version
     register: response
```

##### EXPECTED RESULTS

Should run successfully

##### ACTUAL RESULTS

```
The full traceback is:
Traceback (most recent call last):
  File "/Users/gnalawad/ansible/ganeshrn/ansible/lib/ansible/executor/task_executor.py", line 119, in run
    res = self._execute()
  File "/Users/gnalawad/ansible/ganeshrn/ansible/lib/ansible/executor/task_executor.py", line 523, in _execute
    result = self._handler.run(task_vars=variables)
  File "/Users/gnalawad/ansible/ganeshrn/ansible/lib/ansible/plugins/action/junos.py", line 73, in run
    socket_path = connection.run()
  File "/Users/gnalawad/ansible/ganeshrn/ansible/lib/ansible/plugins/connection/persistent.py", line 68, in run
    socket_path = self._start_connection()
  File "/Users/gnalawad/ansible/ganeshrn/ansible/lib/ansible/plugins/connection/persistent.py", line 96, in _start_connection
    result = json.loads(to_text(stdout, errors='surrogate_then_replace'))
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/json/__init__.py", line 338, in loads
    return _default_decoder.decode(s)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/json/decoder.py", line 369, in decode
    raise ValueError(errmsg("Extra data", s, end, len(s)))
ValueError: Extra data: line 1 column 8 - line 4 column 121 (char 7 - 405)

fatal: [38.145.35.195]: FAILED! => {
    "msg": "Unexpected failure during module execution.",
    "stdout": ""
}
```

### PR DESCRIPTION
**Pull request by ganeshrn**
##### SUMMARY

Fixes #33047

~*  As debug logs are written on stdout, it interrupts
   the communication between ansible-connection(background)
  process and main process. To avoid this add a string delimiter to exactly 
  identify the response string.~

*  ansible-connection which runs as a background process sends a
   json string (contains response received from remote device)
   to foreground ansible-playbook process over stdout.

*  If in case debug flag is enabled the connection_loader api
   invoked from ansible-connection `ssh = connection_loader.get('ssh', class_only=True)`
   results in emitting debug logs on stdout. This  spurious log
   interfere with the actual response and results in failure while
   reading json string in ansible-playbook process

* To avoid this save stdout of ansible-connection and redirect it string
  buffer to accumulate all the logs emitted by core API's

* Add these logs in `result['messages']` which is send a json string after reinstating saved stdout

*  Remove unwanted code in ansible-connection

* Add `stdin.flush()` after write to  handle buffer scenario in case of PY3
 

##### ISSUE TYPE

 - Bugfix Pull Request

##### COMPONENT NAME

bin/ansible-connection
executor/task_executor.py
plugins/connection/persistent.py
##### ANSIBLE VERSION

```
2.5
```

##### ADDITIONAL INFORMATION

```

```

### CODE REVIEW

**Comment by Qalthos** [comment]
This does work around the problem, but loses all the debug info. I think I'd rather seen ansible-connection take over stdout/stderr so those lines can get pushed back as messages properly, rather than just dropping them.

**Comment by ganeshrn** [comment]
I didn't get your comment. AFAIK ansible-connection sends the request to cliconf/netconf plugin over jsonrpc and waits for the response which is in json format. However, while it is waiting a spurious debug log is send over stdout which is also considered as part of the response.
Within `ansible-connection` probably there is no way to differentiate between debug log and json response.

**Comment by ganeshrn** [comment]
~Debug logs can be extrated in `executor/task_executor.py` and copy it to messages.~
~`
logs = out.split(br"##RESPONSE##:")[0]
out = out.split(br"##RESPONSE##:")[1]
`~

With latest commit debug logs are now displayed on the console.

File: lib/ansible/executor/task_executor.py [hunk 1]
```diff
@@ -817,10 +817,12 @@ def _start_connection(self):
...
-        if p.returncode == 0:
-            result = json.loads(stdout)
-        else:
-            result = json.loads(stderr)
+        out = stdout if p.returncode == 0 else stderr
+        if br"##RESPONSE##:" not in out:
+            return None
+
+        out = out.split(br"##RESPONSE##:")[1]
+        result = json.loads(to_text(out.rstrip(), errors='surrogate_then_replace'))
```
Comment (kedarkekan) [inline lib/ansible/executor/task_executor.py]: please use out.strip() to handle leading whitespaces from messages too.

File: lib/ansible/plugins/connection/persistent.py [hunk 2]
```diff
@@ -92,10 +93,12 @@ def _start_connection(self):
...
-        if p.returncode == 0:
-            result = json.loads(to_text(stdout, errors='surrogate_then_replace'))
-        else:
-            result = json.loads(to_text(stderr, errors='surrogate_then_replace'))
+        out = stdout if p.returncode == 0 else stderr
+        if br"##RESPONSE##:" not in out:
+            return None
+
+        out = out.split(br"##RESPONSE##:")[1]
+        result = json.loads(to_text(out.rstrip(), errors='surrogate_then_replace'))
```
Comment (kedarkekan) [inline lib/ansible/plugins/connection/persistent.py]: Let us print the verbose logs as well? 
`for line in output[0].splitlines():`

File: lib/ansible/executor/task_executor.py (hunk 1 above)
Comment (ganeshrn) [reply lib/ansible/executor/task_executor.py]: Fixed

File: lib/ansible/plugins/connection/persistent.py (hunk 2 above)
Comment (ganeshrn) [reply lib/ansible/plugins/connection/persistent.py]: Fixed

**Review by kedarkekan — APPROVED** [review]

**Comment by bdowling** [comment]
Is this ready to be merged?  It addressed the issue for me.  Thanks!

**Comment by ganeshrn** [comment]
@bdowling The approach in the fix is not correct as message passed between background and foreground process is expected to be a json string. Adding a delimiter before json string defeats the purpose.

I am working on an alternate approach to fix this issue in a better way.

**Review by gundalow — APPROVED** [review]

**Comment by ganeshrn** [comment]
ready_for_review

**Pull request merged**